### PR TITLE
Few alternate forms order and rotating ID function

### LINKED ISSUE #2020: PUT THE MINIOR IN RAINBOW ORDER
**Current Behavior**
Okay, so currently I believe the order of the minior cores in the pokedex and pokemon list is in alphabetical order by color, which is:

Blue-core, Green-core, Indigo-core, Orange-core, Red-core, Violet-core, Yellow-core

**Improved Behavior**
The improvement i would love to see would be to change the way this order is displayed in the pokedex and pokemon list to the classic rainbow order of:

Red-core, Orange-core, Yellow-core, Green-core, Blue-core, Indigo-core, Violet-core

**Reason for change**
This has no practical use or reason i just think it will look way better.

**Detailed Description**
Change the order of miniors core colors in the pokedex and pokemon list to the order of the rainbow or the order of the colors in visible light if you wanna get scientific.

### PR DESCRIPTION
**Pull request by osherwott**
Switched order of forms in PokemonList.ts and PokemonNameTypes.ts to accurately reflect official and game listings:
* Pikachu (World Cap) between (Partner) and (Flying)
* Vivillon (Icy Snow) as 666.00, (Meadow) as 666.06, and (Poké Ball) as 666.18
* Flabébé, Floette, and Florges (Orange) between (Yellow) and (Blue)
* Minior in *ROGBIV* order: (Red), (Orange), (Yellow), (Green), (Blue), (Indigo), (Violet)
* Alcremie by cream: (Vanilla), (Ruby Cream), (Matcha), (Mint), (Lemon), (Salted), (Ruby Swirl), (Caramel), (Rainbow), and within each cream subdivided by sweet: (Strawberry), (Berry), (Love), (Star), (Clover), (Flower), (Ribbon); ID subdivisions to two decimal places changed accordingly

Re-named files of images (shiny and non-shiny) to match new IDs:
* Pikachu 25.08 - 25.12
* Vivillon 666 - 666.18
* Flabébé 669.02 and 669.03
* Floette 670.02 and 670.03
* Florges 671.02 and 671.03
* Minior 774.01 - 774.07
* Alcremie 869.01 - 869.86 (tenths and hundredths numbers changed accordingly)

Re-added Pumpkaboo (Average) non-shiny sprite that had been omitted for some time after being replaced with Pumpkaboo (Small)

Adjusted code in PokedexHelper.ts to account for Vivillon (Meadow)'s new ID

Resolves #2020

### CODE REVIEW

**Comment by Farboo171** [comment]
There are a LOT of things that would need to be checked for dependency from this change. Any file that references an alt form (from quests immediately comes to mind) by number would be affected.

**Comment by osherwott** [comment]
I figured that'd be the case. Would you know where the code for individual quests is located? I was not able to see any specific quests in the files I found.

**Comment by Farboo171** [comment]
https://github.com/pokeclicker/pokeclicker/blob/develop/src/scripts/quests/QuestLineHelper.ts

And I am sure there are other places that have dependencies that I don't know about.

Checking all the dependencies for all the Pokedex numbers AND writing the correct Update functions is a massive undertaking, probably best split into a few PRs. Maybe one per mon. And if you ask around in the discord , other people more familiar with other parts of the code can probably give more details on places to check

**Comment by osherwott** [comment]
Thank you. As of now I haven't noticed any issues, it seems that all the forms I reordered use their names as reference instead of ID. I have checked QuestLineHelper.ts, Farming.ts, Discord.ts, and BattleCafe.ts. Are there any other dependencies other than obtain methods I should be aware of?

**Comment by Jaaslet** [comment]
This will need an update function. We changed the id of the event mons in like version 9, so you can look there

**Comment by Aegyo** [comment]
Can you keep this PR to only your flabebe, floette, florges, and minior changes please? The others are being changed around already and this conflicts with some decisions there.

Your update function will need to update the id for pokemon in the party list, and move statistics to their new id as well. Looks like this was done in [0.9.0](https://github.com/pokeclicker/pokeclicker/blob/develop/src/scripts/Update.ts#L598), but there are more statistics to move now - check PokemonHelper.incrementPokemonStatistics

Might be easier to just open a new PR with just the requested pokemon, but I'll leave that decision to you.

**Comment by osherwott** [comment]
Thanks for the information Aegyo. I'll check over the new statistics and figure out what to do with my PR later. If you don't mind me asking, is there anything I could do for the other forms since I'm already working on this? Ideally I would like to help any way I can, but understand if you'd rather work on it yourselves.

**Comment by Aegyo** [comment]
I think the best thing you can do to help would be making your update function a generic static property on Update, like moveIndex and changeHatcheryKey are.

I would call it something like `rotatePokemonIds` and have it take a list of ids that need rotating around.
eg it looks like minior has these rotations `[774.01, 774.05]` `[774.02, 774.04]` `[774.03, 774.06, 774.07]`

**Comment by PixLSteam** [comment]
> I think the best thing you can do to help would be making your update function a generic static property on Update, like moveIndex and changeHatcheryKey are.
> 
> I would call it something like `rotatePokemonIds` and have it take a list of ids that need rotating around. eg it looks like minior has these rotations `[774.01, 774.05]` `[774.02, 774.04]` `[774.03, 774.06, 774.07]`

When doing stuff like that, we have to be careful not to introduce a potential bug in future versions. For example, data model changes could easily break older update scripts if they rely on a static method.

**Comment by Aegyo** [comment]
I think for this it should be fine? If the save data model changes, then we would just need to change the name to `xxxPre0_10_24` (or whatever version) and have a new version to work with the latest data model?

Or am I missing something?

Maybe all of these static functions should be versioned from the start to save effort later, but that can be handled separately

**Comment by PixLSteam** [comment]
I think it should be fine for this as long as we ignore missing statistics. We should probably rename older helpers whenever we change the model, but I'm pretty sure we'd just forget about that so there's no point.

Also, we gotta make sure the rotation works properly. I can't test the current update function but I feel like we'd erase/overwrite a few statistics with what we have right now.

**Comment by osherwott** [comment]
I'm afraid I'm having difficulty with arrays. I'm not sure how to make the function account for different row lengths, nor if I should differentiate the existing function from 0.9.0 in some way. I'm not experienced with coding so this is more of a hobby and it will take me a while to figure out. I will study the existing code and online in the meanwhile. Any help is appreciated @Aegyo

**Comment by Qwertypop04** [comment]
When you removed the changes to images you removed the image changes for Pokémon that are still changing, as well.

**Comment by osherwott** [comment]
Don't worry, I'm aware of that! I have prioritized on the function, but will remember to address the other file changes before the PR is finalized!

**Comment by PixLSteam** [comment]
> I'm afraid I'm having difficulty with arrays. I'm not sure how to make the function account for different row lengths, nor if I should differentiate the existing function from 0.9.0 in some way. I'm not experienced with coding so this is more of a hobby and it will take me a while to figure out. I will study the existing code and online in the meanwhile. Any help is appreciated

Maybe something like this:

```js
const obj = {
	a: 1,
	b: 2,
	c: 3,
	d: 4,
	e: 5,
	f: 6
};
function rotate(rotations) {
	const firstValue = obj[rotations[0]];
	for (let i = 0; i < rotations.length - 1; i++) {
		obj[rotations[i]] = obj[rotations[i + 1]];
	}
	obj[rotations[rotations.length - 1]] = firstValue;
}
```
Keep in mind that this is just a POC and you'd have to figure out how to do this with all of the relevant statistics as well as the PartyPokemon data.

**Comment by Aegyo** [comment]
> if I should differentiate the existing function from 0.9.0
This should be different, because you are moving pokemon around into existing IDs, and won't need to delete their previous id locations (because something else will overwrite there anyway)
The one in 0.9.0 is removing an old id and creating a new one for the data to live in.

**Comment by osherwott** [comment]
Also could this please be marked as a Draft. Thanks

**Comment by Aegyo** [comment]
> Help is needed

What kind of help would you like? We can suggest more code (either the whole thing or for a particular issue you are having), or someone can push changes directly to your pr branch for you to us.

**Comment by osherwott** [comment]
Sorry for the late reply, it turned out the if statements from the copied 0.9.0 function was messing with the function. I believe it should be error free now, however it does not appear to be working. I will change the image and list files for now and go over the function later. Any review or pushes by others would be appreciated in the meanwhile. Thanks

File: src/scripts/Update.ts [hunk 1]
```diff
@@ -1395,6 +1395,19 @@ class Update implements Saveable {
...
+                // Flabebe line (swap Blue and Orange)
+                [669.02, 669.03],
+                [670.02, 670.03],
+                [671.02, 671.03],
+                // ROYGBIV Minior instead of BGIORVY
+                [774.01, 774.05],
+                [774.02, 774.04],
+                [774.03, 774.06, 774.07],
+            ];
+            Update.rotatePokemonIDs(saveData, formIDs);
```
Comment (Aegyo) [inline src/scripts/Update.ts:1410]: Had a quick look through, your rotatePokemonIDs looks fine, but this line I think should be:
```ts
        formIDs.forEach((list) => Update.rotatePokemonIDs(saveData, list));
```

On mobile so I can't test it out right now though.

**Comment by osherwott** [comment]
I've been trying with a save file for hours and it doesn't work. If your dummy data went through then I have no idea why it won't work on the game itself. Upon further testing it seems that I can't even overwrite any data at all. I hope you are having better luck.

**Comment by Aegyo** [comment]
Looks like we've put the update in the v0.10.1 section, try moving it over to the v0.10.2 bit and load a save that hasn't been on v0.10.2 yet :sweat_smile:

**Comment by osherwott** [comment]
I just noticed you moved the statistics in #3374, could that be affecting it?

**Comment by Aegyo** [comment]
I don't think so, but I'm looking into it now

**Comment by Aegyo** [comment]
Had some capitalization errors, seems to work with an actual save file after fixing those and moving into the v0.10.2 section.
Do you want me to sort out the conflicts too?

**Comment by osherwott** [comment]
> Had some capitalization errors
Does this mean the original 0.9.0 function should be rewritten with capital letters as well? It might not work currently, and it'd be best for posterity.
> Do you want me to sort out the conflicts too?
Yes please, and thank you!

Also PixLSteam mentioned PartyPokemon data. I suppose this means held items, nicknames, and the like. I'm not sure how to include these in this function, since there wasn't such code in the 0.9.0 one. I want to avoid causing any bugs with this function if used in the future, especially since mega stones are being currently implemented.

**Comment by PixLSteam** [comment]
All the data directly associated with the PartyPokemon should be copied over since we just change the id property - even if we add new properties in the future.

**Comment by osherwott** [comment]
It works!

**Comment by osherwott** [comment]
Oh, but before it's merged, can I rotate the Vivillon? #3371 didn't put Poke Ball pattern towards the end

**Comment by osherwott** [comment]
Ready to merge!

**Comment by osherwott** [comment]
Also @Aegyo should I make a new PR for a "move" function? Now that we have this working we can trim it down and add "delete old statistics" to help Qwertypop04 move all those IDs in #3371

**Comment by Aegyo** [comment]
No need imo, it can be handled in Qwerty's PR

**Review by Jaaslet — CHANGES_REQUESTED** [review]
Some of the images have been changed in the docs folder instead of the src folder. All changes should be in the src folder.
It's a bit hard for me to test when the images are wrong, so i will test the rest, when that has been fixed :)

**Comment by Aegyo** [comment]
@osherwott do you think you can get those images sorted in the next 6-ish hours, before Jaaslet comes back for another review?

Non-shiny images for 669, 670, 671, and 774 variants were made in the docs folder instead of src.
Also looks like 666.10 didn't get an image update, but it is in the rotation list. The image would be named 666.1.png, since that 666.10 is treated as a number and loses the trailing 0.

**Comment by osherwott** [comment]
Sorry for the delay! Everything should be fine now.

**Comment by Aegyo** [comment]
Image updates for 666.10 still missing, unless I'm being blind again :sweat_smile:

**Comment by osherwott** [comment]
What a headache! _NOW_ everything should be fine!

**Comment by Aegyo** [comment]
Looks ok to me too, hopefully Jaaslet agrees :crossed_fingers:

**Review by Jaaslet — APPROVED** [review]
Looks good. Thanks for the PR

**Pull request merged**